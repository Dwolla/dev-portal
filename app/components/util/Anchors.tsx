import {
  Children,
  cloneElement,
  createContext,
  useContext,
  useEffect,
  useState,
} from "react";
import { useRouter } from "next/router";
import GithubSlugger from "github-slugger";
import scrollToElement from "scroll-to-element";
import { childrenToString } from "../../modules/helpers";

// constants

const ANCHOR_EL_TYPES = ["h1", "h2", "h3"];
export const SCROLL_OFFSET = 100;
export const SCROLL_DURATION = 500;

// AnchorsContext

export const AnchorsContext = createContext({
  anchors: [] as Anchor[],
  setAnchors: (_anchors: Anchor[]) => {}, // eslint-disable-line no-unused-vars
  activeAnchor: null as Anchor,
  setActiveAnchor: (anchor: Anchor) => {}, // eslint-disable-line no-unused-vars
});

export const useAnchors = () => useContext(AnchorsContext);

// AnchorsProvider

export function AnchorsProvider(props: {
  children: JSX.Element | JSX.Element[];
}) {
  const [anchors, setAnchors] = useState([]);
  const [activeAnchor, setActiveAnchor] = useState(null);

  return (
    <AnchorsContext.Provider
      value={{
        anchors,
        setAnchors,
        activeAnchor,
        setActiveAnchor,
      }}
    >
      {props.children}
    </AnchorsContext.Provider>
  );
}

// AnchorsSetter

function AnchorData(props: { anchor: Anchor; children: JSX.Element }) {
  return props.children;
}

export function AnchorsSetter(props) {
  const { setAnchors, activeAnchor, setActiveAnchor } = useAnchors();

  const slugger = new GithubSlugger();

  const children =
    Children.map(props.children, (c) => {
      if (ANCHOR_EL_TYPES.includes(c.props.originalType)) {
        const text = childrenToString(c.props.children);
        const id = slugger.slug(text);
        const level = ANCHOR_EL_TYPES.indexOf(c.props.originalType) + 1;
        return (
          <AnchorData anchor={{ id, text, level }}>
            {cloneElement(c, { id })}
          </AnchorData>
        );
      }
      return c;
    }) || [];

  const anchors = children.reduce(
    (acc, current) =>
      current.props.anchor ? [...acc, current.props.anchor] : acc,
    []
  );

  const router = useRouter();

  useEffect(() => {
    setAnchors(anchors);
    setActiveAnchor(anchors[0]);

    return () => {
      setAnchors([]);
      setActiveAnchor(null);
    };
  }, [router?.pathname]);

  useEffect(() => {
    let raf: number;
    const anchorEls = anchors.map((a: Anchor) => document.getElementById(a.id));

    const handleScroll = () => {
      raf = requestAnimationFrame(() => {
        if (anchorEls && anchorEls.length > 0) {
          const anchorId = anchorEls.reduce(
            (acc: { score: number; el: { id: string } }, next) => {
              const { y } = next.getBoundingClientRect();
              const nextScore = Math.abs(SCROLL_OFFSET - y);

              if (!acc) {
                return { el: next, score: nextScore };
              }

              return nextScore < acc.score
                ? { el: next, score: nextScore }
                : acc;
            },
            null
          ).el.id;

          if (anchorId !== activeAnchor?.id) {
            setActiveAnchor(anchors.find((a: Anchor) => a.id === anchorId));
          }
        }
      });
    };

    window.addEventListener("scroll", handleScroll);
    window.addEventListener("resize", handleScroll);

    return () => {
      window.removeEventListener("scroll", handleScroll);
      window.removeEventListener("resize", handleScroll);
      cancelAnimationFrame(raf);
    };
  }, [router?.pathname]);

  useEffect(() => {
    if (activeAnchor) {
      window.history.replaceState(
        null,
        null,
        activeAnchor.level !== 1 ? `#${activeAnchor.id}` : " " //prevent h1 tags from displaying as anchors in the url
      );
    }
  }, [activeAnchor?.id]);

  return children;
}

export const scrollTo = (anchorId: string) => {
  const element = document.getElementById(anchorId);

  if (element) {
    scrollToElement(element, {
      offset: -SCROLL_OFFSET + 1,
      duration: SCROLL_DURATION,
    });
  }
};
