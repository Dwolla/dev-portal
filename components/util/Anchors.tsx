import {
  createContext,
  useState,
  useEffect,
  useContext,
  Children,
  cloneElement,
} from "react";
import { useRouter } from "next/router";
import GithubSlugger from "github-slugger";
import { childrenToString } from "../../modules/helpers";

// constants

const ANCHOR_EL_TYPES = ["h1", "h2", "h3", "h4", "h5", "h6"];

// AnchorsContext

export const AnchorsContext = createContext({
  anchors: [] as Anchor[],
  setAnchors: (_anchors: Anchor[]) => {}, // eslint-disable-line no-unused-vars
});

export const useAnchors = () => useContext(AnchorsContext);

// AnchorsProvider

export function AnchorsProvider(props: {
  children: JSX.Element | JSX.Element[];
}) {
  const [anchors, setAnchors] = useState([]);

  return (
    <AnchorsContext.Provider value={{ anchors, setAnchors }}>
      {props.children}
    </AnchorsContext.Provider>
  );
}

// AnchorsSetter

function AnchorData(props: { anchor: Anchor; children: JSX.Element }) {
  return props.children;
}

export function AnchorsSetter(props) {
  const { setAnchors } = useAnchors();

  const slugger = new GithubSlugger();

  const children = Children.map(props.children, (c) => {
    if (ANCHOR_EL_TYPES.includes(c.props.originalType)) {
      const text = childrenToString(c.props.children);
      const id = slugger.slug(text);
      const level = ANCHOR_EL_TYPES.indexOf(c.props.originalType) + 1;
      return (
        <AnchorData anchor={{ id, text, level }}>
          {cloneElement(c, { id })}
        </AnchorData>
      );
    }
    return c;
  });

  const anchors = children.reduce(
    (acc, current) =>
      current.props.anchor ? [...acc, current.props.anchor] : acc,
    []
  );

  const router = useRouter();

  useEffect(() => {
    setAnchors(anchors);

    return () => setAnchors([]);
  }, [router?.pathname]);

  return children;
}
