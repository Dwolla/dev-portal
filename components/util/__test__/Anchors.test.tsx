import { create as render, act } from "react-test-renderer";
import { createElement } from "react";
import { useAnchors, AnchorsProvider, AnchorsSetter } from "../Anchors";

const mockPath = "/";
jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: mockPath,
      pathname: mockPath,
      query: "",
      asPath: mockPath,
    };
  },
}));

function DebugAnchors() {
  return <pre>{JSON.stringify(useAnchors().anchors, null, 2)}</pre>;
}

function MockMDXCreateElement(props: {
  children: string;
  originalType: string;
  id?: string;
}) {
  return createElement(props.originalType, {
    children: props.children,
    id: props.id,
  });
}

test("Anchors", async () => {
  let tree;
  act(() => {
    tree = render(
      <AnchorsProvider>
        <AnchorsSetter>
          <MockMDXCreateElement originalType="h1">A</MockMDXCreateElement>
          <MockMDXCreateElement originalType="h2">B</MockMDXCreateElement>
          <MockMDXCreateElement originalType="h3">C</MockMDXCreateElement>
          <MockMDXCreateElement originalType="h4">D</MockMDXCreateElement>
          <MockMDXCreateElement originalType="h5">E</MockMDXCreateElement>
          <MockMDXCreateElement originalType="h6">F</MockMDXCreateElement>
          <MockMDXCreateElement originalType="h1">A</MockMDXCreateElement>
          <MockMDXCreateElement originalType="h2">A</MockMDXCreateElement>
        </AnchorsSetter>

        <DebugAnchors />
      </AnchorsProvider>
    );
  });

  expect(tree.toJSON()).toMatchSnapshot();
});
