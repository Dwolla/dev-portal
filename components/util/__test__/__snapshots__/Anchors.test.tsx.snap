// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchors 1`] = `
Array [
  <h1
    id="a"
  >
    A
  </h1>,
  <h2
    id="b"
  >
    B
  </h2>,
  <h3
    id="c"
  >
    C
  </h3>,
  <h4
    id="d"
  >
    D
  </h4>,
  <h5
    id="e"
  >
    E
  </h5>,
  <h6
    id="f"
  >
    F
  </h6>,
  <h1
    id="a-1"
  >
    A
  </h1>,
  <h2
    id="a-2"
  >
    A
  </h2>,
  <pre>
    [
  {
    "id": "a",
    "text": "A",
    "level": 1
  },
  {
    "id": "b",
    "text": "B",
    "level": 2
  },
  {
    "id": "c",
    "text": "C",
    "level": 3
  },
  {
    "id": "d",
    "text": "D",
    "level": 4
  },
  {
    "id": "e",
    "text": "E",
    "level": 5
  },
  {
    "id": "f",
    "text": "F",
    "level": 6
  },
  {
    "id": "a-1",
    "text": "A",
    "level": 1
  },
  {
    "id": "a-2",
    "text": "A",
    "level": 2
  }
]
  </pre>,
]
`;
